File: apps/chat/libs/chat-storage/src/main/java/com/example/chat/storage/mapper/FriendshipMapper.java
package com.example.chat.storage.mapper;

import org.springframework.stereotype.Component;

import com.example.chat.domain.friendship.Friendship;
import com.example.chat.domain.friendship.FriendshipId;
import com.example.chat.domain.user.UserId;
import com.example.chat.storage.entity.ChatFriendshipEntity;

/**
 * Friendship Domain ↔ ChatFriendshipEntity 변환
 */
@Component
public class FriendshipMapper {

    /**
     * Domain → Entity 변환
     */
    public ChatFriendshipEntity toEntity(Friendship friendship) {
        return ChatFriendshipEntity.builder()
                .id(friendship.getId().value())
                .userId(friendship.getUserId().value())
                .friendId(friendship.getFriendId().value())
                .status(friendship.getStatus())
                .nickname(friendship.getNickname())
                .favorite(friendship.isFavorite())
                .createdAt(friendship.getCreatedAt())
                .updatedAt(friendship.getUpdatedAt())
                .build();
    }

    /**
     * Entity → Domain 변환
     */
    public Friendship toDomain(ChatFriendshipEntity entity) {
        return Friendship.fromStorage(
                FriendshipId.of(entity.getId()),
                UserId.of(entity.getUserId()),
                UserId.of(entity.getFriendId()),
                entity.getCreatedAt(),
                entity.getStatus(),
                entity.getNickname(),
                entity.isFavorite(),
                entity.getUpdatedAt());
    }
}

File: apps/chat/libs/chat-storage/src/main/kotlin/com/example/chat/storage/mapper/FriendshipMapper.kt
package com.example.chat.storage.mapper

import com.example.chat.domain.friendship.Friendship
import com.example.chat.domain.friendship.FriendshipId
import com.example.chat.domain.user.UserId
import com.example.chat.storage.entity.ChatFriendshipEntity
import org.springframework.stereotype.Component

/**
 * Friendship Domain ↔ ChatFriendshipEntity 변환
 */
@Component
class FriendshipMapper {

	/**
	 * Domain → Entity 변환
	 */
	fun toEntity(friendship: Friendship): ChatFriendshipEntity {
		return ChatFriendshipEntity(
			id = friendship.id.value,
			userId = friendship.userId.value,
			friendId = friendship.friendId.value,
			status = friendship.status,
			nickname = friendship.nickname,
			favorite = friendship.favorite,
			createdAt = friendship.createdAt,
			updatedAt = friendship.updatedAt
		)
	}

	/**
	 * Entity → Domain 변환
	 */
	fun toDomain(entity: ChatFriendshipEntity): Friendship {
		return Friendship(
			id = FriendshipId.of(entity.id),
			userId = UserId.of(entity.userId),
			friendId = UserId.of(entity.friendId),
			status = entity.status,
			nickname = entity.nickname,
			favorite = entity.favorite,
			createdAt = entity.createdAt,
			updatedAt = entity.updatedAt
		)
	}
}

/**
 * Extension Functions for more idiomatic Kotlin
 */
fun Friendship.toEntity(): ChatFriendshipEntity {
	return ChatFriendshipEntity(
		id = id.value,
		userId = userId.value,
		friendId = friendId.value,
		status = status,
		nickname = nickname,
		favorite = favorite,
		createdAt = createdAt,
		updatedAt = updatedAt
	)
}

fun ChatFriendshipEntity.toDomain(): Friendship {
	return Friendship(
		id = FriendshipId.of(id),
		userId = UserId.of(userId),
		friendId = UserId.of(friendId),
		status = status,
		nickname = nickname,
		favorite = favorite,
		createdAt = createdAt,
		updatedAt = updatedAt
	)
}
